### PR TITLE
feat!: indexing rework

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Completely rework the indexing API, with tests

### CODE REVIEW

File: docs/src/api.md [hunk 1]
```diff
@@ -1,20 +1,13 @@
...
-param_sym_to_index
-is_param_sym
+isvariable
+variableindex
+isparameter
+parameterindex
+isindependent_variable
+isobserved
+observed
+istimedependent
```
Comment (oscardssmith) [inline docs/src/api.md:11]: the naming/underscore placement here seems slightly inconsistent

Comment (AayushSabharwal) [reply docs/src/api.md:11]: `isindependentvariable` feels unreadable to me. Maybe `isindepvar`?

File: src/SymbolicIndexingInterface.jl [hunk 2]
```diff
@@ -1,13 +1,104 @@
...
+"""
+    issymbolic(x) = issymbolic(typeof(x))
+    issymbolic(::Type)
+
+Check whether a type implements the [`Symbolic`](@ref) trait or not. Default to
+[`NotSymbolic`](@ref) for all types except `Symbol`.
+"""
+issymbolic(x) = issymbolic(typeof(x))
+issymbolic(::Type) = NotSymbolic()
+issymbolic(::Type{Symbol}) = Symbolic()
```
Comment (YingboMa) [inline src/SymbolicIndexingInterface.jl:36]: Is there more possibilities than `true/false` from `issymbolic`?

Comment (AayushSabharwal) [reply src/SymbolicIndexingInterface.jl:36]: I was considering adding a `SymbolicArray` trait variant, for `SymArr` and `AbstractArray{<:something that implements Symbolic}` but it's just as easy to check `all(isequal(Symbolic()), issymbolic.(collect(sym)))`

File: docs/src/api.md (hunk 1 above)
Comment (AayushSabharwal) [reply docs/src/api.md:11]: Fixed in https://github.com/SciML/SymbolicIndexingInterface.jl/pull/12/commits/9bea16466126a7c57c8305c11597aea8beb190b7

**Review by YingboMa — APPROVED** [review]

**Review by oscardssmith — APPROVED** [review]

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
🎉

**Pull request merged**